README.md

                             mbrChunker v0.3.17
                              by JD Durick 
                           <labgeek@gmail.com>
                          <http://vmforensics.org>

               Licensed under the GNU General Public License v2

General
---------- 
mbrChunker is a *nix-based program that allows you to mount raw disk images (created by dd, dcfldd, dc3dd, ftk imager, etc) by taking the
 raw image, analyzing the master boot record (physical sector 0), and getting specific information that is needed to create a vmdk file.
 Version 0.3.17 has the ability to extract data from the master boot record of any raw image, extracted information such as heads
 cylinders, sectors per track, etc. and create a working vmdk file (monolithic flat disk).  The tool now has the ability to search 
 for hex byte patterns within any type of binary files (ie.  dd images of VMFS partitions). 
 
 Compile:
 1.  Download mbrChunker-0.3.17.tar.gz (tar zxvf mbrChunker-0.3.17.tar.gz)
 2.  cd mbrChunker-0.3.17
 3.  make
 
root@redbox:/data/projects/C/mbrChunker-0.3.17# make
gcc -g -Wall -o mbrChunker mbrChunker.c dd2vmdk.c mbr.c
  
 4.  run mbrChunker:  ./mbrChunker (to see options)
 
 
 Below is your options for using mbrChunker:
 
root@redbox:/data/projects/C/mbrChunker-0.3-release# ./mbrChunker -h
Program: mbrChunker
Author:  JD Durick <jd@vmforensics.org>
Website: http://vmforensics.org

Description:  mbrChunker is a *nix-based program that allows you to mount raw disk images
as well as perform analysis on partition tables within the master boot record.  Furthermore,
mbrChunker can search for hex patterns within any type of binary file
Usage: ./mbrChunker -i <path to dd image> -v <output path of vmdk>...

OPTIONS:
   -i image file name that you will be analyzing
   -v VMDK metadata file that you will be creating for your dd2vmdk conversion
   -a creation of a metadata file to be created (file location required)
   -x search for hex code patterns within your file (not 0x80, just 80)
   -b blocksize you will be using during your hex pattern search [allows for huge files to be searched]
   
Examples:
./mbrChunker -i <raw image path location> -x <hex pattern> -b <blocksize for buffer>
./mbrChunker -i <raw image path location> -v <vmdk output file location>
./mbrChunker -i <raw image path location> -a <output of analysis file>
 
 

Features
-----------
For all *nix users:
    o Converts raw dd images to vmdk files
    o handles large files (over the 2GB limit)
    o Quick and easy to use
    o Metadata file gives you additional information about the image file you are converting/analzying
    o searches for hex strings within a dd image or any binary file
    o matches number of hex strings within large file (<4gb)
    o For VMFS partitions that have been dd'd:  searches for all MBR of all VM's within the dd image
    	 - after it grabs the hex offset for each string, then gets the full 512 master boot record 
    	   and then parses it.  This applies to VMFS partitions so far.
    o With the -x and -i features, you can now find hex byte offsets within your dd image or file.  I 
      was tired of using hexedit(great tool) so I figured I would integrate it into the mbrChunker tool.
      This allows the user to enter the hex pattern via the command line such as hex bytes:  4939614747 
      as well as your image, it will then find the byte location within that image.  I thought it was useful 
      to me.  What I will end up doing is going through a VMFS partition and finding the MBR partitions
      and grabbing all 512 bytes, analyzing them, then printing out those results.  We also have a -b option
      that allows the user to predefine the buffer block size for which you are searching.  This allow you to 
      search files greater than 4GB in size.
    

Limitation(s):
 - The DD to vmdk functionality only handles images that use Partition #1 (not dual booted) or using
 any of the other primary partition tables.
 - lack of error checking throughout
 
Usage
------

./mbrChunker -i <raw image path location> -v <vmdk output file location>
./mbrChunker -i <raw image path location> -a <output of analysis file>
./mbrChunker -x 4939614747 -i /data/images/testsector.img 

TODO
----
 - clean up

EXAMPLE #1
---------
root@redbox:/data/projects/C/mbrChunker/Debug# ./mbrChunker -i /data/images/sixgbimage.img -v /tmp/output.vmdk

Author:   JD Durick <jd@vmforensics.org>
Program:  mbrChunker
Website:  http://vmforensics.org
Version:  0.2
Revision date:  2010-07-05 08:27:47 -0400 (Mon, 05 Jul 2010)

Image Geometry specifications [/data/images/sixgbimage.img]
================================
1.  Image location:  /data/images/sixgbimage.img
1.  VMDK destination:  /tmp/output.vmdk
2.  Number of sectors = 12594960
3.  Number of cylinders:  1023
4.  Number of heads per track:  64
5.  Number of sectors per track: 63
6.  File size in bytes: 6448619520
7.  Size of each sector: 512

EXAMPLE #2
----------
root@redbox:/data/projects/C/mbrChunker/Debug# ./mbrChunker -i /data/images/firstsector.img -a /tmp/output_analysis.txt

Author:   JD Durick <jd@vmforensics.org>
Program:  mbrChunker
Website:  http://vmforensics.org
Version:  0.2
Revision date:  2010-07-05 08:27:47 -0400 (Mon, 05 Jul 2010)
======================================

MBR Analysis of /data/images/firstsector.img
Metadata file location:  /tmp/output_analysis.txt
File size in bytes:         512

PARTITION 1:
============

Partition 1 has an active partition:  0x80
Starting Head:	0x01 (1)
System ID:	0x07
Partition 1 (VALID):  07
Ending head:	0xfe (254)
Sectors between MBR and Partition 1:  63
Total sectors in partition 1:  429690492

PARTITION 2:
============

Partition 2 does not have an active partition:  0x00
Starting Head:	0xfe (254)
System ID:	0x0f
Partition 2 (VALID):  0f
Ending head:	0xfe (254)
Sectors between MBR and Partition 2:  36539
Total sectors in partition 2:  547077510

PARTITION 3:
============

Partition 3 does not have an active partition:  0x00
Starting Head:	0x00 (0)
System ID:	0x00
*** Unused partition table entry ***
Partition 3 (VALID):  00
Ending head:	0x00 (0)
Sectors between MBR and Partition 3:  0
Total sectors in partition 3:  0

PARTITION 4:
============

Partition 4 does not have an active partition:  0x00
Starting Head:	0x00 (0)
System ID:	0x00
*** Unused partition table entry ***
Partition 4 (VALID):  00
Ending head:	0x00 (0)
Sectors between MBR and Partition 4:  0
Total sectors in partition 4:  0

Two byte signature word:  55AA (End of MBR)
Exiting...

EXAMPLE #3
----------
In the following example, from the hex signature offset, I grab the next 5 bytes - however, 
I will enable that later until I get more time to work on it.  For now, you will be just getting the 
byte offset location:

root@redbox:/data/projects/C/mbrChunker-0.3.17# ./mbrChunker -i /data/images/image57.img -x 3961 -b 25

Total size of /data/images/image57.img file:  57 bytes
Parsing Hex string "3961" now.

Results:

Hex string 3961 was found at starting byte location:  2
Hex string 3961 was found at starting byte location:  17
Hex string 3961 was found at starting byte location:  33
Hex string 3961 was found at starting byte location:  52


Quick statistics:
=====================
Number of bytes that have been read:  57
Number of signature matches found:  4
Total number of bytes in hex string:  2
Block size of each buffer:  25
Number of buffers searched:  3

Requirements
------------
GCC:
root@redbox:/data/projects/C/mbrChunker-0.3-release# gcc -v
Using built-in specs.
Target: i486-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.3.3-5ubuntu4' --with-bugurl=file:///usr/share/doc/gcc-4.3/README.Bugs --enable-languages=c,c++,fortran,objc,obj-c++ --prefix=/usr --enable-shared --with-system-zlib --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --enable-nls --with-gxx-include-dir=/usr/include/c++/4.3 --program-suffix=-4.3 --enable-clocale=gnu --enable-libstdcxx-debug --enable-objc-gc --enable-mpfr --enable-targets=all --with-tune=generic --enable-checking=release --build=i486-linux-gnu --host=i486-linux-gnu --target=i486-linux-gnu
Thread model: posix
gcc version 4.3.3 (Ubuntu 4.3.3-5ubuntu4) 

               


CHANGELOG
/* SVN FILE: $Id: CHANGELOG 5 2010-07-14 02:26:52Z jd $*/

Version 0.1
* first working version
* minimal error checking
* lacking efficiency and extended functionality

Version 0.2
* added functionality to analyze all 4 primary partitions
* added some modular features

Version 0.2.1
*  can now parse each of the primary partition entries, 
   analyze each of the 16 byte entries for information such as boot indicator, starting head, 
   starting sector, starting cylinder, system ID, ending head, ending sector, ending cylinder, and total sectors
*  adds a analyzed function plus can output analysis to file

Version 0.3.17
* added hex byte offset searching function which allows you to search for a hexadecimal
  string within a dd image file (VMFS partition)
* modularized
* 

dd2vmdk.c
/* SVN FILE: $Id: dd2vmdk.c 5 2010-07-14 02:26:52Z jd $*/
/*
 * Source file : Program to parse the MBR of a dd image.
 * $Author: jd $
 * $Date: 2010-07-13 22:26:52 -0400 (Tue, 13 Jul 2010) $
 * $Revision: 5 $
 * $LastChangedBy: jd $
 * $URL: file:///data/codehole/mbrChunker/trunk/dd2vmdk.c $
 */

#include <stdio.h>

const int mstr_boot_record = 512;

int getcylinders(int p1[]) {

	int begincylinder, endcyclinder, cylinders;
	/*
	 * starting cylinder is a 10 bit value,
	 * taking 8 bits from partition entry
	 *  and 2 bits from partition entry 3
	 */
	begincylinder = ((p1[3] << 8) | p1[2]);
	endcyclinder = ((p1[7] << 8) | p1[6]);

	/*
	 * code from PartitionEntry.java
	 * in the Liveview source code
	 */
	cylinders = ((endcyclinder & 65280) >> 8);

	// if the 6th bit of the 16 bit
	// structure is not zero
	if ((endcyclinder & 64) != 0) {
		cylinders += 256;
	}
	// set bit 8 of the 10 bit cylinder value

	if ((endcyclinder & 128) != 0) { // if the 7th bit of the 16 bit
		// structure is not zero
		cylinders += mstr_boot_record; // set bit 9 of the 10 bit cylinder value
	}

	/* loop until null is found */
	//for (n = 0; n < partitionentrysize; n++) {
	//	printf("%d and %02X\n", n, p1[n]);
	//}
	//printf("\n\n");

	return cylinders;

}

int getendcylinder(int partition1[]) {
	int endcyl = 0;
	endcyl = ((partition1[7] << 8) | partition1[6]);

	return endcyl;
}

int getsectorspertrack(int endcyl) {
	int sectors_per_track = 0;
	/* the ending sector is 6 bits which would be 00111111
	 * use the bitwise operator AND and compares in binary form
	 * in the case of 65535,  = 11111111 11111111
	 * Comparing the two would give us still 63 decimal
	 */
	sectors_per_track = endcyl & 63;
	return sectors_per_track;

}
int getheadspertrack(int part1[]) {

	int nheads = 0;
	nheads = part1[1] + part1[5];

	return nheads;

}

/*
 * Compliments to zapotek's template for creating a
 * baseline VMDK metadata file
 *
 */

void printvmdkfile(char *src, char *output, long tsectors, int cyls, int heads,
		int sectpertrack) {

	FILE *outputfile;
	outputfile = fopen(output, "w");
	fprintf(outputfile, "\nversion=1\n");
	fprintf(outputfile, "encoding=\"UTF-8\"\n");
	fprintf(outputfile, "CID=fffffffe\n");
	fprintf(outputfile, "parentCID=ffffffff\n");
	fprintf(outputfile, "isNativeSnapshot=\"no\"\n");
	fprintf(outputfile, "createType=\"monolithicFlat\"\n\n");
	fprintf(outputfile, "RW %ld FLAT \"%s\" 0\n\n", tsectors, src);
	fprintf(outputfile, "ddb.virtualHWVersion = \"7\"\n");
	fprintf(outputfile,
			"ddb.longContentID = \"29075898903f9855853610dffffffffe\"\n");
	fprintf(outputfile,
			"ddb.uuid = \"60 00 C2 91 8e 73 27 62-43 58 3b f8 05 ae 2e a0\"\n");
	fprintf(outputfile, "ddb.geometry.cylinders = \"%d\"\n", cyls);
	fprintf(outputfile, "ddb.geometry.heads = \"%d\"\n", heads);
	fprintf(outputfile, "ddb.geometry.sectors = \"%d\"\n", sectpertrack);
	fprintf(outputfile, "ddb.adapterType = \"ide\"\n");
	fclose(outputfile);

}



mbrChunker.h
/* SVN FILE: $Id: mbrChunker.h 19 2010-07-30 00:52:44Z jd $*/
/*
 * Project Name : mbrChunker.h header file
 * $Author: jd $
 * $Date: 2010-07-29 20:52:44 -0400 (Thu, 29 Jul 2010) $
 * $Revision: 19 $
 * $LastChangedBy: jd $
 * $URL: file:$
 */

#ifndef MBRCHUNKER_H_
#define MBRCHUNKER_H_

#define PROGRAM_NAME "mbrChunker"
#define AUTHOR "JD Durick <jd@vmforensics.org>"
#define WEBSITE "http://vmforensics.org"
#define VERSION "0.3 $Rev: 19 $"
#define DEFAULT_BLOCKSIZE 1024
#define HEXSTRING_MAX 16
#define HEX_SIZE 256

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <ctype.h>
#include <inttypes.h>
#include <errno.h>

/*
 * Function prototypes
 */

off_t getfilesize(FILE*fp);
void usage(char *);
void getpe1(int[]);
int getendcylinder(int[]);
int getsectorspertrack(int);
int getcylinders(int[]);
int getheadspertrack(int[]);
void printvmdkfile(char *, char *, long, int, int, int);
void parsePartitionEntries(char *masterbootrec, FILE *);
void mbrAnalysis(FILE *, char *, char *, long long);
char *getMBR(FILE *);
char* dec2hex(int);
int validatePartitionType(char *, FILE *);
int analyzePartitions(int[], int, FILE *);
void end_of_sector_marker(char *, FILE *);
void hxsearch(FILE *, char *, long long, long);
char *allocateBuffer(FILE *fptr, long long);

#endif /* MBRCHUNKER_H_ */



mbr.c
/* SVN FILE: $Id: mbr.c 5 2010-07-14 02:26:52Z jd $*/
/*
 * Source file : Program to parse the MBR of a dd image.
 * $Author: jd $
 * $Date: 2010-07-13 22:26:52 -0400 (Tue, 13 Jul 2010) $
 * $Revision: 5 $
 * $LastChangedBy: jd $
 * $URL: file:///data/codehole/mbrChunker/trunk/mbr.c $
 */

#include "mbrChunker.h"
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>

#define REVISIONDATE "2010-07-05 08:27:47 -0400 (Mon, 05 Jul 2010)"

/*
 * listing of the partition types similar to when you toggle while using
 * fdisk.  Pulled this from Tim Vidas and Brian Kaplan's Liveview java source code.
 */

const int validPartitionTypes[] = { 0x00, 0x10, 0x02, 0x03, 0x04, 0x05, 0x06,
		0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x14,
		0x16, 0x17, 0x18, 0x1b, 0x1c, 0x1e, 0x24, 0x39, 0x3c, 0x40, 0x41, 0x42,
		0x4d, 0x4e, 0x4f, 0x50, 0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x5c, 0x61,
		0x63, 0x64, 0x65, 0x70, 0x75, 0x80, 0x81, 0x82, 0x83, 0x84, 0x85, 0x86,
		0x87, 0x88, 0x8e, 0x93, 0x94, 0x9f, 0xa0, 0xa5, 0xa6, 0xa7, 0xa8, 0xa9,
		0xab, 0xb7, 0xb8, 0xbb, 0xbe, 0xbf, 0xc1, 0xc4, 0xc6, 0xc7, 0xda, 0xdb,
		0xde, 0xdf, 0xe1, 0xe3, 0xe4, 0xeb, 0xee, 0xef, 0xf0, 0xf1, 0xf4, 0xf2,
		0xfd, 0xfe, 0xff };
//const int startingByte = 0;
//const int partitionentrysize = 16;
//const int bootcodesize = 446;
/*
 * Input:  File pointer and source file pointer
 * Output:  boolean success or failure
 * Desc:   Attempts to open the file and read it,
 *         passes, back file pointer.
 */

void mbrAnalysis(FILE *f, char *sourcefile, char *output, long long size) {

	char *masterbootrecord;
	FILE *mdata_output_file;
	mdata_output_file = fopen(output, "w");

		printf("\nAuthor:   %s\n", AUTHOR);
		printf("Program:  %s\n", PROGRAM_NAME);
		printf("Website:  %s\n", WEBSITE);
		printf("Version:  %s\n", VERSION);
		printf("Revision date:  %s\n", REVISIONDATE);
		printf("\nMBR Analysis of %s\n", sourcefile);
		printf("Metadata file location:  %6s\n", output);
		printf("File size in bytes:  %lld\n", size);

		fprintf(mdata_output_file, "\nAuthor:   %s\n", AUTHOR);
		fprintf(mdata_output_file, "Program:  %s\n", PROGRAM_NAME);
		fprintf(mdata_output_file, "Website:  %s\n", WEBSITE);
		fprintf(mdata_output_file, "Version:  %s\n", VERSION);
		fprintf(mdata_output_file, "Revision date:  %s\n", REVISIONDATE);
		fprintf(mdata_output_file, "======================================\n");

		fprintf(mdata_output_file, "\nMBR Analysis of %s\n", sourcefile);
		fprintf(mdata_output_file, "Metadata file location:  %6s\n", output);
		fprintf(mdata_output_file, "File size in bytes:  %lld\n", size);

		/*
		 * Grabs the master boot record.  We pass
		 * the File pointer in first.
		 */
		masterbootrecord = getMBR(f);
		parsePartitionEntries(masterbootrecord, mdata_output_file);
		fclose(mdata_output_file);
		return;



}


/*
 * Gets the master boot record (in bytes)
 * input:  FILE pointer to opened file
 * output:  pointer to master boot record
 */

char *getMBR(FILE *fptr) {
	char *bfr;
	int bytes;

	bfr = malloc(512);
	if (bfr == NULL) {
		printf("Error, malloc failed, wtf\n");
		exit(EXIT_FAILURE);
	}

	(fseek(fptr, 0, SEEK_SET));
	bytes = fread(bfr, 1, 512, fptr);
	return (bfr);

}

void parsePartitionEntries(char *masterbootrec, FILE *outputptr) {

	int partitionentry1[16] = { };
	int partitionentry2[16] = { };
	int partitionentry3[16] = { };
	int partitionentry4[16] = { };
	int ptnum1 = 1;
	int ptnum2 = 2;
	int ptnum3 = 3;
	int ptnum4 = 4;

	int rtn = 0;
	char *hexvalue_p1;
	int i = 0;
	int counter = -1;

	for (i = 446; i < 446 + 16; i++) {
		counter++;
		hexvalue_p1 = dec2hex(masterbootrec[i] & 0xff);
		//printf("partition 1:  %d --> %02d ---> hex[%s]\n", counter,
		//		masterbootrec[i] & 0xff, hexvalue_p1);
		partitionentry1[counter] = masterbootrec[i] & 0xff;
		partitionentry2[counter] = masterbootrec[i + 16] & 0xff;
		partitionentry3[counter] = masterbootrec[i + 32] & 0xff;
		partitionentry4[counter] = masterbootrec[i + 48] & 0xff;

	}

	rtn = analyzePartitions(partitionentry1, ptnum1, outputptr);
	rtn = analyzePartitions(partitionentry2, ptnum2, outputptr);
	rtn = analyzePartitions(partitionentry3, ptnum3, outputptr);
	rtn = analyzePartitions(partitionentry4, ptnum4, outputptr);
	end_of_sector_marker(masterbootrec, outputptr);

	return;

}

int analyzePartitions(int p1[], int pnumber, FILE *oPtr) {

	fprintf(oPtr, "\nPARTITION %d:\n", pnumber);
	fprintf(oPtr, "============\n");
	int i = 0;
	int sectors_mbr_partition = 0;
	int isValid = 0;
	char *hexvalue = NULL;
	char *starthead, *systemid, *endhead;
	long total_sectors = 0;

	for (i = 0; i < 16; i++) {
		//printf("[%d] --> %d\n", i, p1[i]);
	}

	hexvalue = dec2hex(p1[0] & 0xff);
	if (p1[0] == 128) {
		fprintf(oPtr, "\nPartition %d has an active partition:  0x%s\n",
				pnumber, hexvalue);

	} else {
		fprintf(oPtr,
				"\nPartition %d does not have an active partition:  0x%s\n",
				pnumber, hexvalue);
	}

	/*
	 * get your starting head location
	 */
	starthead = dec2hex(p1[1]);
	fprintf(oPtr, "Starting Head:\t0x%s (%d)\n", starthead, p1[1]);

	/*
	 * get your system ID
	 * compare this with the array of
	 * valid partitions
	 */
	systemid = dec2hex(p1[4]);
	fprintf(oPtr, "System ID:\t0x%s\n", systemid);

	isValid = validatePartitionType(systemid, oPtr);
	if (isValid) {
		fprintf(oPtr, "Partition %d (VALID):  %s\n", pnumber, systemid);
	} else {
		fprintf(oPtr, "Partition %d (NOT VALID):  %s\n", pnumber, systemid);
	}

	endhead = dec2hex(p1[5]);
	fprintf(oPtr, "Ending head:\t0x%s (%d)\n", endhead, p1[5]);

	sectors_mbr_partition = (p1[9] << 8 | p1[8]);
	fprintf(oPtr, "Sectors between MBR and Partition %d:  %d\n", pnumber,
			sectors_mbr_partition);

	total_sectors = (p1[15] << 24 | p1[14] << 16 | p1[13] << 8 | p1[12]);
	fprintf(oPtr, "Total sectors in partition %d:  %ld\n", pnumber,
			total_sectors);

	return 0;

}

int validatePartitionType(char *code, FILE *out) {

	int i = 0;
	char *hx;
	char *unused = "00";

	for (i = 0; i < 94; i++) {
		hx = dec2hex(validPartitionTypes[i]);

		if (strcmp(hx, code) == 0) {
			if (strcmp(code, unused) == 0) {
				fprintf(out, "*** Unused partition table entry ***\n");
			}
			//printf("hx = %s and code = %s\n", hx, code);
			return 1;
		}
	}
	// if you got here, no match.
	return 0;

}
